meson.build

```diff
@@ -0,0 +1,800 @@
+    error('host compiler `@0@` doesn\'t match host architecture `@1@`'.format(
+        host_machine.cpu_family(), get_option('arch')))
+endif
+
+################################
+# additional compilation flags #
+################################
+
+if get_option('debugging')
+    common_cflags += ['-O0', '-g']
+else
+    common_cflags += ['-O2', '-g']
+endif
+
+if get_option('trace')
+    trace_cflags       = ['-pg', '-fno-omit-frame-pointer']
+    uftrace_cflags    += trace_cflags
+    demangler_cflags  += trace_cflags
+    symbols_cflags    += trace_cflags
+    dbginfo_cflags    += trace_cflags
+    traceevent_cflags += trace_cflags
+    test_cflags       += trace_cflags
+    # cannot add -pg to libmcount_cflags because mcount() is not reentrant
+endif
+
+if get_option('asan')
+    asan_cflags = ['-O0', '-g', '-fsanitize=address,leak']
+    common_ldflags   += ['-fsanitize=address,leak']
+
+    uftrace_cflags    += asan_cflags
+    demangler_cflags  += asan_cflags
+    symbols_cflags    += asan_cflags
+    dbginfo_cflags    += asan_cflags
+    traceevent_cflags += asan_cflags
+    test_cflags       += asan_cflags
+endif
+
+if get_option('coverage')
+libmcount_cflags
+    coverage_cflags   = ['-O0', '-g', '--coverage', '-U_FORTIFY_SOURCE']
+    common_ldflags   += ['--coverage']
+    common_cflags    += coverage_cflags
+    libmcount_cflags += coverage_cflags
+    test_cflags      += coverage_cflags
+endif
+
+if get_option('san') != ''
+    if get_option('san') == 'all'
+        sanitize_cflags = ['-O0', '-g', '-fsanitize=address,leak,undefined']
+        common_ldflags += ['-fsanitize=address,leak,undefined']
+    else
+        sanitize_cflags = ['-O0', '-g', '-fsanitize=@0@'.format(get_option('sanitize'))]
+        common_ldflags += ['-fsanitize=@0@'.format(get_option('sanitize'))]
+    endif
+
+    uftrace_cflags    += sanitize_cflags
+    demangler_cflags  += sanitize_cflags
+    symbols_cflags    += sanitize_cflags
+    dbginfo_cflags    += sanitize_cflags
+    traceevent_cflags += sanitize_cflags
+    test_cflags       += sanitize_cflags
+endif
+
+################################
+# -Dwith-capstone build option #
+################################
+
+libcapstone_is_enabled = false
+libcapstone_dep = dependency('capstone', required: get_option('with-capstone'))
+if libcapstone_dep.found()
+    libcapstone_is_enabled = true
+    common_cflags += '-DHAVE_LIBCAPSTONE'
+endif
+
+#############################
+# -Dwith-libdw build option #
+#############################
+
+libdw_is_enabled = false
+libdw_dep = []
+if not get_option('with-libelf').disabled()
+    libdw_dep = cc.find_library('dw', required: get_option('with-libdw'))
+    if libdw_dep.found()
+        libdw_is_enabled = true
+        common_cflags += '-DHAVE_LIBDW'
+    endif
+endif
+
+##############################
+# -Dwith-libelf build option #
+##############################
+
+libelf_is_enabled = false
+libelf_dep = cc.find_library('elf', required: get_option('with-libelf'))
```

Comment:
By default, it does not check for headers. But it is pretty easy by adding the `has_headers` argument, as specified in the [documentation](https://mesonbuild.com/Reference-manual.html#compiler-object).

By `grep`-ing in the code, `gelf.h`, `libelf.h` and `elfutils/libdw.h` seem required. I'll add them.

Should I also do the same for the other dependencies?

Reply:
I think so.  Btw how does it check the dependency?  Is it possible to use pkg-config to check the availability and to list needed flags?

Reply:
I'm not sure how `cc.find_library()` works, I believe it does a test compile.

But I searched a little. Turns out I'm using `dependency` for other dependencies rather than `cc.find_library()`. The `dependency()` function will use `auto` method by default. It may choose `pkg-config` or `cmake` to find library. See the [docs](https://mesonbuild.com/Dependencies.html#dependency-method).

The `dependency()` can't verify for header existance. I don't think it is necessary is the dependency is found using `pkg-config`. If the library has a `pkg-config` it means it has been installed for development purpose and headers should be present.

Still, I'll change `libelf` and `libdw` to `dependency()` as it seems more robust.